worlds/pokemon_rb/__init__.py

```diff
@@ -209,196 +225,378 @@ def encode_name(name, t):
+                intervene_move = "Cut"
+            elif ((not logic.can_learn_hm(test_state, "Flash", self.player)) and self.multiworld.dark_rock_tunnel_logic[self.player]
+                    and (((self.multiworld.accessibility[self.player] != "minimal" and
+                    (self.multiworld.trainersanity[self.player] or self.multiworld.extra_key_items[self.player])) or
+                    self.multiworld.door_shuffle[self.player]))):
+                intervene_move = "Flash"
+            # If no Pokémon can learn Fly, then during door shuffle it would simply not treat the free fly maps
+            # as reachable, and if on no door shuffle or simple, fly is simply never necessary.
+            # We only intervene if a Pokémon is able to learn fly but none are reachable, as that would have been
+            # considered in door shuffle.
+            elif ((not logic.can_learn_hm(test_state, "Fly", self.player)) and logic.can_learn_hm(test_state, "Fly", self.player)
+                    and self.multiworld.door_shuffle[self.player] not in
+                    ("off", "simple") and [self.fly_map, self.town_map_fly_map] != ["Pallet Town", "Pallet Town"]):
+                intervene_move = "Fly"
             if intervene_move:
                 if intervene_move == last_intervene:
                     raise Exception(f"Caught in infinite loop attempting to ensure {intervene_move} is available to player {self.player}")
-                intervene(intervene_move)
+                intervene(intervene_move, test_state)
                 last_intervene = intervene_move
             else:
                 break
 
+        # Delete evolution events for Pokémon that are not in logic in an all_state so that accessibility check does not
+        # fail. Re-use test_state from previous final loop.
+        evolutions_region = self.multiworld.get_region("Evolution", self.player)
+        clear_cache = False
+        for location in evolutions_region.locations.copy():
+            if not test_state.can_reach(location, player=self.player):
+                evolutions_region.locations.remove(location)
+                clear_cache = True
+        if clear_cache:
+            self.multiworld.clear_location_cache()
+
+
         if self.multiworld.old_man[self.player] == "early_parcel":
             self.multiworld.local_early_items[self.player]["Oak's Parcel"] = 1
             if self.multiworld.dexsanity[self.player]:
-                for location in [self.multiworld.get_location(f"Pokedex - {mon}", self.player)
-                                 for mon in poke_data.pokemon_data.keys()]:
-                    add_item_rule(location, lambda item: item.name != "Oak's Parcel" or item.player != self.player)
+                for mon in poke_data.pokemon_data:
+                    try:
+                        location = self.multiworld.get_location(f"Pokedex - {mon}", self.player)
```

Comment:
multiworld.get_location causes a full recache of all locations in the entire multiworld on cache miss. Which mons exist should be tracked in a way that doesn't cause this to occur.

Reply:
(There are several instances of this, I will not mark each in this review=

Reply:
Done